### PR TITLE
fixed race is_idle

### PR DESCRIPTION
**Pull request by obi458**
## Overview

Fixed the race in is_idle

## Testing recommendations

Race is difficult to test

## Checklist

- [x] Code is written and works correctly;
- [ ] Changes are covered by tests;
- [ ] Documentation reflects the changes;

### CODE REVIEW

**Comment by janl** [comment]
Fix for #1181

File: src/couch/src/couch_db.erl [hunk 1]
```diff
@@ -200,7 +200,12 @@ reopen(#db{main_pid = Pid, fd = Fd, fd_monitor = OldRef, user_ctx = UserCtx}) ->
 
 incref(#db{fd = Fd} = Db) ->
     Ref = erlang:monitor(process, Fd),
-    {ok, Db#db{fd_monitor = Ref}}.
+    receive
+        {'DOWN', Ref, _, _, _} ->
```
Comment (davisp) [inline src/couch/src/couch_db.erl]: You should pattern match the reason element as noproc so we're asserting that the process was dead before we tried to acquire the monitor. I'd also swap the return value to be {ok, Db} on success, or retry on failure. No need to return the input and pattern match it on use.

File: src/couch/src/couch_server.erl [hunk 2]
```diff
@@ -77,16 +77,20 @@ open(DbName, Options0) ->
     case ets:lookup(couch_dbs, DbName) of
     [#entry{db = Db0, lock = Lock} = Entry] when Lock =/= locked ->
         update_lru(DbName, Entry#entry.db_options),
-        {ok, Db1} = couch_db:incref(Db0),
-        couch_db:set_user_ctx(Db1, Ctx);
+        case couch_db:incref(Db0) of
+            {down,Db0} -> open(DbName, Options0);
```
Comment (davisp) [inline src/couch/src/couch_server.erl]: You should add a Retry argument to this function so that we limit ourself to a defined number of attempts or else this could lead to infinite looping (granted that's unlikely but its a possibility if we don't prevent it).

Comment (davisp) [reply src/couch/src/couch_server.erl]: And we should also do the same retry limiting for create as well.

File: src/couch/src/couch_db.erl (hunk 1 above)
Comment (davisp) [reply src/couch/src/couch_db.erl]: Thinking more on this, I'd use the atom down instead of retry. Was thinking in terms of the couch_server's response, but that's not appropriate when considering incref more abstractly. So something like `-spec incref(Db::db()) -> {ok, NewDb::db()} | down.`

**Comment by obi458** [comment]
After some times on the couch and thinking about the problem i guess is_idle is wrong using monitored_by.

I simulate scheduler(pause) always with an debug point!

If i set an debug point on:

Process(Fd):
```
handle_info(maybe_close, File) ->
    case is_idle(File) of
        true ->
         **debug point ->**  {stop, normal, File};
        false ->
            erlang:send_after(?MONITOR_CHECK, self(), maybe_close),
            {noreply, File}
    end;
```

and calling erlang:monitor(process, Fd) from process(OpenDb) that want to open Db:

```
receive
        {'DOWN', Ref, _, _, _} ->
            {down, Db}
    after 0 ->
        {ok, Db#db{fd_monitor = Ref}}
    end.
```

does not work, cause the Fd process lives(on my debug point)! 
and will die shortly through {stop, normal, File}

I think it is better to replace the monitor function with(and turn around the logic):

gen_server:call(Fd,{monitored_by,OpenDb},unlimited).
if the calling failed, we are sure the process does not exists and we restart to open or create the Db.

In short: 

1) save the OpenDb in monitored_by dict and 
2) the couch_file monitors the OpenDb and remove OpenDb from monitored_by dict on 'DOWN' message.
     Therefore we have always a consistent monitored_by dict and is_idle will work correctly!

is_idle like:
is_idle() -> dict:size(monitored_by_dict) =:= 0.

**Comment by davisp** [comment]
I thought about something similar the other day however I didn't go too far down that route as the cost of a message pass through couch_file would be quite a performance penalty especially as the system load increases.

I'm starting to wonder though if there's not an alternative approach to having couch_file just notify couch_server that its idle. Though that suggests extra messaging through couch_server that might also be a performance issue on the top end.

**Comment by obi458** [comment]
One solution may be:

If we are going to close the Db(couch_file) we try:

```
is_idle() ->
....
andalso ets:insert_new(couch_dbs,self()).
```


Further if we are going to monitor the Fd:

 ```
incref(#db{fd = Fd} = Db) ->
  incref(Db,ets:insert_new(couch_dbs,Fd)).

incref(#db{fd = Fd} = Db, true = _Ready) ->
    % first monitor
    Ref = erlang:monitor(process, Fd),
    % than release the lock
    ets:delete(couch_dbs,Fd),
    {ok, Db#db{fd_monitor = Ref}};
% we are not able to write the lock,
% either the file will be closed or another process want to monitor
% we will try it again
incref(#db{fd = _Fd} = Db, false = _Ready) ->
    retry.
```

and 

```
...
case couch_db:incref(Db0) of
                retry -> open(DbName, Options0);
                {ok, Db1} -> couch_db:set_user_ctx(Db1, Ctx)
            end;
...

...
case couch_db:incref(Db0) of
                retry -> create(DbName, Options0);
                {ok, Db1} -> couch_db:set_user_ctx(Db1, Ctx)
            end;
...

```

**Comment by davisp** [comment]
Aha! I was reading back through history and staring at this code to remind myself how this fits together and I realized that this isn't actually the right bug. The issue here is actually that we're doing a dirty read on the ets table and then attempting to take the monitor in an unprotected section. The proposed issues we've had here won't ever actually help us (other than the retry being a possible bandaid that'd likely fix it well enough).

The dirty read issue becomes apparent when I finally remembered that couch_db_updater always monitors couch_file. So what must have happened is that couch_db_updater sent the close_db_if_idle cast which gets processed just after a client reads the db record from the ets table. What should really happen is that the monitor is taken while the client has locked the db entry in the couch_dbs table. Which is similar-ish to the proposal to insert/delete pids from that table.

However, actually doing that makes things *really* slow (I've tried it for other reasons). And also fairly complicated as you then have the possibility where clients die during a "transaction" which then leaves your table with permanent locks and so on and such forth.

I've run across this for some other reason I can't remember as well. And I've never been able to figure out a fix that doesn't cripple the throughput of couch_server.

**Comment by obi458** [comment]
The motivation we thinking about the problem is: a lot of noproc errors in changes!
Following by a lot of timeouts!
Maybe we could better handle the badmatch!

```
[error] 2018-02-23T18:55:04.251497Z couchdb@127.0.0.1 <0.12749.108> 8647c35089 rexi_server: from: couchdb@127.0.0.1(<0.960.81>) mfa: fabric_rpc:changes/3 error:{badmatch,{'EXIT',noproc}} [{couch_file,pread_binary,2,[{file,"src/couch_file.erl"},{line,169}]},{couch_file,pread_term,2,[{file,"src/couch_file.erl"},{line,157}]},{couch_btree,get_node,2,[{file,"src/couch_btree.erl"},{line,434}]},{couch_btree,reduce_stream_node,11,[{file,"src/couch_btree.erl"},{line,619}]},{couch_btree,fold_reduce,4,[{file,"src/couch_btree.erl"},{line,81}]},{couch_db,count_changes_since,2,[{file,"src/couch_db.erl"},{line,1407}]},{fabric_rpc,changes,4,[{file,"src/fabric_rpc.erl"},{line,80}]},{rexi_server,init_p,3,[{file,"src/rexi_server.erl"},{line,139}]}]
[error] 2018-02-23T19:11:26.204435Z couchdb@127.0.0.1 <0.12625.108> 8647c35089 rexi_server: from: couchdb@127.0.0.1(<0.960.81>) mfa: fabric_rpc:changes/3 exit:timeout [{rexi,init_stream,1,[{file,"src/rexi.erl"},{line,256}]},{rexi,stream_last,2,[{file,"src/rexi.erl"},{line,224}]},{fabric_rpc,changes,4,[{file,"src/fabric_rpc.erl"},{line,86}]},{rexi_server,init_p,3,[{file,"src/rexi_server.erl"},{line,139}]}]
[error] 2018-02-23T19:11:26.206475Z couchdb@127.0.0.1 <0.12799.108> 8647c35089 rexi_server: from: couchdb@127.0.0.1(<0.960.81>) mfa: fabric_rpc:changes/3 exit:timeout [{rexi,init_stream,1,[{file,"src/rexi.erl"},{line,256}]},{rexi,stream_last,2,[{file,"src/rexi.erl"},{line,224}]},{fabric_rpc,changes,4,[{file,"src/fabric_rpc.erl"},{line,86}]},{rexi_server,init_p,3,[{file,"src/rexi_server.erl"},{line,139}]}]
[error] 2018-02-23T19:11:26.206597Z couchdb@127.0.0.1 <0.12936.108> 8647c35089 rexi_server: from: couchdb@127.0.0.1(<0.960.81>) mfa: fabric_rpc:changes/3 exit:timeout [{rexi,init_stream,1,[{file,"src/rexi.erl"},{line,256}]},{rexi,stream_last,2,[{file,"src/rexi.erl"},{line,224}]},{fabric_rpc,changes,4,[{file,"src/fabric_rpc.erl"},{line,86}]},{rexi_server,init_p,3,[{file,"src/rexi_server.erl"},{line,139}]}]
```
This comes from:

ioq:
```
submit_request(#request{}=Request, #state{}=State) ->
    Ref = erlang:monitor(process, Request#request.fd), % fd is dead!!!
```

```
handle_info({'DOWN', Ref, _, _, Reason}, State) ->
    case lists:keytake(Ref, #request.ref, State#state.running) of
        {value, Request, Remaining} ->
            gen_server:reply(Request#request.from, {'EXIT', Reason}), % Reason is noproc
            {noreply, State#state{running=Remaining}, 0};
        false ->
            {noreply, State, 0}
    end;
```

**Comment by obi458** [comment]
I don't give up :-)

```
incref(#db{fd = Fd} = Db) ->
    incref(Db,couch_file:lock(Fd)).

incref(#db{fd = Fd} = Db, true = _Locked) ->
    % first monitor
    Ref = erlang:monitor(process, Fd),
    % than release the lock
    couch_file:unlock(Fd),
    {ok, Db#db{fd_monitor = Ref}};
% we are not able to write the lock,
% either the file will be closed or another process want to monitor
% we will try it again
incref(#db{fd = _Fd} = _Db, false = _Locked) ->
    retry.
```

couch_file:

```
% System dbs aren't monitored by couch_stats_process_tracker
is_idle(#file{is_sys=true}) ->
    case process_info(self(), monitored_by) of
        {monitored_by, []} -> lock(self());
        _ -> false
    end;
is_idle(#file{is_sys=false}) ->
    Tracker = whereis(couch_stats_process_tracker),
    case process_info(self(), monitored_by) of
        {monitored_by, []} -> lock(self());
        {monitored_by, [Tracker]} -> lock(self());
        {monitored_by, [_]} -> exit(tracker_monitoring_failed);
        _ -> false
    end.

lock(Fd) -> try
                ets:new(list_to_atom(pid_to_list(Fd)), [named_table]), true
            catch
                _:_ -> false
            end.
unlock(Fd) -> ets:delete(list_to_atom(pid_to_list(Fd))).
```

couchd_db_updater:

```
handle_info(timeout, #db{fd=Fd, name=DbName} = Db) ->
    IdleLimitMSec = update_idle_limit_from_config(),
    case couch_db:is_idle(Db) of
        true ->
            % we have to unlock Fd, cause close_db_if_idle will check is_idle again
            couch_file:unlock(Fd),
            MSecSinceLastRead = couch_file:msec_since_last_read(Fd),
            case MSecSinceLastRead > IdleLimitMSec of
                true ->
                    ok = couch_server:close_db_if_idle(DbName);
                false ->
                    ok
            end;
        false ->
            ok
    end,
    % Send a message to wake up and then hibernate. Hibernation here is done to
    % force a thorough garbage collection.
    gen_server:cast(self(), wakeup),
    {noreply, Db, hibernate}.
```

Process died -> ets died and lock will released!

`timer:tc(fun() -> [begin couch_file:lock(self()), couch_file:unlock() end || _ <- lists:seq(1,1000000)] end).`

{2191728,ok} on iMac Pro

**Comment by janl** [comment]
@obi458 would you be able to resolve the conflicts?

**Comment by wohali** [comment]
@davisp can you review this one again please?

**Comment by wohali** [comment]
Let's try to get this into 2.3.0. That means you need to resolve the conflicts @obi458 and we need another review from @davisp .

**Comment by davisp** [comment]
Will try and get a review done this week.

File: src/couch/src/couch_file.erl [hunk 3]
```diff
@@ -702,18 +703,25 @@ monitored_by_pids() ->
...
-        [] -> true;
-        [Tracker] -> true;
+        [] -> lock(self());
+        [Tracker] -> lock(self());
         [_] -> exit(tracker_monitoring_failed);
         _ -> false
     end.
 
+lock(Fd) -> try
+                ets:new(list_to_atom(pid_to_list(Fd)), [named_table]), true
```
Comment (davisp) [inline src/couch/src/couch_file.erl]: This doesn't pass my smell test. Using ets tables as locks has got to be asking for trouble. Between the system limit on tables and no guaranteed cleanup that's a bunch of bad behaviors just waiting to happen.

Comment (obi458) [reply src/couch/src/couch_file.erl]: I pulled a new one using register, without ets!

Comment (davisp) [reply src/couch/src/couch_file.erl]: While register does solve some of those issues that's gonna take a significant amount of work to characterize the performance. I don't think it was ever intended for this sort of usage. I'd like to see some fairly extensive benchmarking before even considering accepting this and also probably some fairly detailed writeup of the register internals explaining how this doesn't become a global lock across shard lookups.

**Comment by wohali** [comment]
@obi458 Any plans to get this in? I am preparing to mass-close old PRs that never got merged.

**Comment by davisp** [comment]
You can close this. If/when we bump to Erlang 21 as a minimum version we could try and revive this using the atomics module.

**Comment by nickva** [comment]
With Erlang 23+ we do have access to atomics so perhaps this could revived using that. That idea would be to keep an atomic ref in the couch_server `entry...` and then http://www.erlang.org/doc/man/atomics.html#compare_exchange-4 to handle the decref/incref locks and retries.
